carbon-design-system/carbon
# fix(combobox): memoize floating ui ref to prevent max call depth

Closes #16857

This PR does to primary things:
1. Memoize the value of `getMenuProps` to avoid an infinite loop
2. Only apply floating ui refs when `autoAlign={true}`

The first fixes the problem outlined in the original issue, the second is an indirect improvement. If the refs had initially been guarded, this bug would've been isolated to users using `autoAlign`. Because the refs were not guarded, the bug was present for everyone regardless of if `autoAlign` was being used.

From what I can tell, there is not a reliable way to unit test this as the keyboard entry rate isn't configurable.

#### Changelog

**Changed**

- Wrap `getMenuProps` in `useMemo`
- Conditionally apply refs

#### Testing / Reviewing

- Go to the default combobox story. Type super fast - like just totally spam the keyboard. The component should not error
- Do the same on the `allowCustomInput` combobox story
- Do the same on the `with experimental autoalign` combobox story
- Ensure `autoAlign` works as expected on combobox by resizing the browser window, playing with scroll, and opening/closing the listbox
